File: C++/BOJ/BOJ1325.cpp
#include <cstdio>
#include <vector>
using namespace std;

void dfs(int current, int &count, vector<vector<int>> &graph, vector<bool> &isVisited, int m) {
    count++;
    isVisited[current] = true;
    for (int idx = 0; idx < graph[current].size(); idx++) {
        if (!isVisited[graph[current][idx]]) {
            dfs(graph[current][idx], count, graph, isVisited, m);
        }
    }
}

int main() {
    int n, m;
    scanf("%d %d", &n, &m);
    vector<vector<int>> graph(n + 1, vector<int>());

    for (int i = 0; i < m; i++) {
        int a, b;
        scanf("%d %d", &a, &b);
        graph[b].push_back(a);
    }

    vector<int> answer(n + 1, 0);
    int maxCount = 0;

    for (int start = 1; start <= n; start++) {
        int count = 0;
        vector<bool> isVisited(n + 1, false);
        dfs(start, count, graph, isVisited, m);
        answer[start] = count;
        maxCount = count > maxCount ? count : maxCount;
    }

    for (int idx = 1; idx < answer.size(); idx++) {
        if (answer[idx] == maxCount) {
            printf("%d ", idx);
        }
    }

    return 0;
}

File: swift/boj/boj1325.swift
struct Queue {
    var queue = [Int]()
    var ptr = 0
    var isEmpty: Bool {
        ptr >= queue.count
    }
    mutating func push(_ v: Int) {
        queue.append(v)
    }
    mutating func pop() -> Int {
        let firstValue = queue[ptr]
        ptr += 1
        if ptr > 50000 {
            queue = Array(queue[ptr...])
            ptr = 0
        }
        return firstValue
    }
}

func bfs(_ start: Int) -> Int {
    var queue = Queue()
    var isVisited = [Bool](repeating: false, count: n + 1)
    queue.push(start)
    isVisited[start] = true
    var linkCount = 0

    while !queue.isEmpty {
        let current = queue.pop()
        for next in graph[current, default: []] {
            if isVisited[next] {
                continue
            }
            isVisited[next] = true
            linkCount += 1
            queue.push(next)
        }
    }
    return linkCount
}

let nm = readLine()!.split(separator: " ").map { Int($0)! }
let (n, m) = (nm[0], nm[1])
var graph: [Int: [Int]] = [:]

for _ in 0..<m {
    let ab = readLine()!.split(separator: " ").map { Int($0)! }
    let (a, b) = (ab[0], ab[1])
    graph[b, default: []].append(a)
}

var answer = [Int]()
var maxCount = 0

for root in 1...n {
    let count = bfs(root)
    if maxCount < count {
        maxCount = count
        answer = [root]
    } else if maxCount == count {
        answer.append(root)
    }
}

for idx in 0..<answer.count {
    print(answer[idx], terminator: idx == answer.count - 1 ? "\n" : " ")
}
